### PR TITLE
Fix and close #621

### LINKED ISSUE #621: Security Question on get_next_url
In scenarios where the `Referer` is not used to control the redirect, users can provide the `next` or `next_url` parameter to specify the next url to visit (redirect).

During testing If I provide an arbitrary or off-site link in one of these parameters to my signup/signin routes, the user gets authenticated and will then be redirected to the value in one of those parameters. The issue is that my OAuth token is copied to the `Referer` header and then sent along to an arbitrary site. An attack scenario is when an attacker supplies a malicious link in the parameter and sends this to a victim. The victim will then click this and see that it goes to the login page for a legitimate site. Upon successful login, they will be redirected to the attacker's site and the authentication token will be leaked and stored on the access logs in the attacker's server. 

**Request**
```
GET /signin/gae/?​next=https://www.attacker.com​ HTTP/1.1
Host: foo.appspot.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:51.0) Gecko/20100101Firefox/51.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Cookie: SACSID=<.. REMOVED ..>
Connection: close
Upgrade-Insecure-Requests: 1
```
**Response**
```
GET / HTTP/1.1
Host: www.attacker.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:51.0) Gecko/20100101 Firefox/51.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Referer: https://www.google.com/url?sa=D&q=https://foo.appspot.com/api/auth/callback/gae/&usg=AFQjCNFaiY<.. REMOVED ..>Q5Z40iMxag
Cookie: B=<...REMOVED...>
Connection: close
Upgrade-Insecure-Requests: 1
```
Is there a way to prevent this?

Also, the `get_next_url` method accepts several protocols such as `javascript:`. Is there a way to only allow `http` and `https` protocols?

### PR DESCRIPTION
**Pull request by gmist**

### CODE REVIEW

File: main/util.py [hunk 1]
```diff
@@ -35,6 +36,17 @@ def switch(case):
...
 
+def is_trusted_url(next_url):
+  if config.TRUSTED_HOSTS:
+    next_url_host = urlparse(next_url).hostname
+    if next_url_host not in config.TRUSTED_HOSTS:
+      return flask.url_for('welcome')
+  else:
+    if not next_url.startswith(flask.request.host_url):
+      return flask.url_for('welcome')
+  return next_url
```
Comment (lipis) [inline main/util.py]:
```python
def is_trusted_url(next_url):
  next_url_host = urlparse(next_url).hostname
  if config.TRUSTED_HOSTS and next_url_host not in config.TRUSTED_HOSTS:
    return flask.url_for('welcome')
  if not next_url.startswith(flask.request.host_url):
    return flask.url_for('welcome')
  return next_url
```

File: main/control/admin.py [hunk 2]
```diff
@@ -54,6 +54,7 @@ class ConfigUpdateForm(flask_wtf.FlaskForm):
   recaptcha_private_key = wtforms.StringField(model.Config.recaptcha_private_key._verbose_name, filters=[util.strip_filter])
   recaptcha_public_key = wtforms.StringField(model.Config.recaptcha_public_key._verbose_name, filters=[util.strip_filter])
   salt = wtforms.StringField(model.Config.salt._verbose_name, [wtforms.validators.optional()], filters=[util.strip_filter])
+  trusted_hosts = wtforms.StringField(model.Config.trusted_hosts._verbose_name, [wtforms.validators.optional()], description='Comma separated: 127.0.0.1, my-domain.com, etc')
```
Comment (lipis) [inline main/control/admin.py]: Comma separated: 127.0.0.1, example.com, etc

File: main/control/admin.py [hunk 3]
```diff
@@ -72,6 +78,10 @@ def admin_config():
     reload(config)
     app.config.update(CONFIG_DB=config_db)
     return flask.redirect(flask.url_for('admin'))
+  else:
```
Comment (lipis) [inline main/control/admin.py]: no need for `else` with `return`

**Comment by lipis** [comment]
ready?

**Comment by gmist** [comment]
It seems to be ready...

**Pull request merged**